shpc/main/modules/templates/no_default_version.tcl

```diff
@@ -0,0 +1,2 @@
+#%Module
+set ModulesVersion "please_specify_a_version_number"
```

Comment:
No, that was not the intent. This is just a way of preventing loading a module without specifying the version number, like what Lmod does by default.
What you're suggesting (a way of defining for each software what the default version should be, instead of letting Lmod/TCL do it based on its own rules) is a different feature, that imho should be spec'ed and addressed in a different PR

Reply:
It works this way (TCL):
```
$ module load samtools/=
ERROR: Unable to locate a modulefile for 'samtools/please_specify_a_version_number'
```

comparing to Lmod:
```
$ module load samtools
Lmod has detected the following error:  The following module(s) are unknown: "samtools"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore_cache load "samtools"

Also make sure that all modulefiles written in TCL start with the string #%Module
```

Reply:
I don't think I fully understand then - could you show me before and after for each (with the PR)? 

And regardless, I'd like to refactor the template generation function in the ways I described because there is a lot of redundancy.

Reply:
Currently, as the documentation says, `default_version` is only implemented for Lmod. It seems its intent is to force people to specify which version they want to load (when set to `false`), and to allow `module load` to arbitrarily choose one version (when set to `true`).

If I've got two `samtools` installed, without my change, `module load samtools` will arbitrarily pick up one of the versions, like Lmod does when the `.version` file is there (i.e. `default_version` set to `true).

What's confusing is that the Lmod implementation needs a `.version` file when `default_version` is `true`, and TCL needs a `.version` file when `default_version` is `false`.

Reply:
> Currently, as the documentation says, default_version is only implemented for Lmod. It seems its intent is to force people to specify which version they want to load (when set to false), and to allow module load to arbitrarily choose one version (when set to true).

Yes I think that's accurate - when default version is false the user has to be explicit, and when true we get lmod to choose for us.

> What's confusing is that the Lmod implementation needs an empty .version file when default_version is true, and nothing when default_version is false. The TCL needs a non-empty .version file when default_version is false, and nothing when default_version is true.

I don't think that is true for the current implementation, but the changes that are in these PRs no? For the current, we just generate .version for lmod and ignore tcl. If this is the only way to make it work, I don't see why the implementation details matter - the user just turns on default versions or not.

So could you:

1. Show me the before/after for running the command with tcl (this is your setup, correct?) I want to see how it responds to `please_specify_a_version_number` because if it gets angry at you, I want shpc to set a reasonable default to be the last installed (and allow the installer to tweak as necessary, or enable/disable based on a setting).
2. Please refactor the function to write the version file to not be redundant.

Reply:
Give me a few minutes, I'll see if I can help you with a few of the changes I'm asking for. I think it might be more clear if I can show you, and I appreciate the work you've already put into this.

Reply:
okay please see https://github.com/muffato/singularity-hpc/pull/1! Ping @marcodelapierre if you want to take a look too.

Reply:
> I don't think that is true for the current implementation, but the changes that are in these PRs no? For the current, we just generate .version for lmod and ignore tcl. If this is the only way to make it work, I don't see why the implementation details matter - the user just turns on default versions or not.

No, that was right. shpc's `default_version` setting is currently irrelevant for TCL, because `default_version==false` is ignored. It's like `default_version` is considered to be true regardless of what the user wants. In this PR, I make TCL refuse to load a software without its version number if there are multiple versions available and `default_version` is set to false.
 
> 1. Show me the before/after for running the command with tcl (this is your setup, correct?) I want to see how it responds to `please_specify_a_version_number` because if it gets angry at you, I want shpc to set a reasonable default to be the last installed (and allow the installer to tweak as necessary, or enable/disable based on a setting).

I've shown how it works in https://github.com/singularityhub/singularity-hpc/pull/517/files#r830613035 . Without the `.version` file, TCL chooses a version and loads it. With a `.version` file, it tries to load the particular version that is in, and complains if the version is not found. This PR assumes that no version would ever be named 'please_specify_a_version_number`.  I don't understand what else I should be showing.

Reply:
@muffato you are responding to the comments / requests I made before I created a PR to your branch. The comments here are now deprecated in response to that.